wp-includes/class-wp-service-workers.php

```diff
@@ -125,6 +150,56 @@ public function init() {
 		do_action_ref_array( 'wp_default_service_workers', array( &$this ) );
 	}
 
+	/**
+	 * Configure offline page based on the saved offline page value.
+	 *
+	 * @return string Script.
+	 */
+	protected function configure_offline_page() {
+		$offline_page_id = (int) get_option( WP_Offline_Page::OPTION_NAME, 0 );
+
+		$replacements = array(
+			'OFFLINE_PAGE_URL' => wp_json_encode( $this->remove_url_scheme( get_the_permalink( $offline_page_id ) ) ),
+		);
+
+		$script = file_get_contents( PWA_PLUGIN_DIR . '/wp-includes/js/offline-page-handling.template.js' ); // phpcs:ignore
+
+		return str_replace(
+			array_keys( $replacements ),
+			array_values( $replacements ),
+			$script
+		);
+	}
+
+	/**
+	 * Get routes for images found in the content of offline post and for the featured image.
+	 *
+	 * @param WP_Post $post Post object.
+	 * @return array Array of routes.
+	 */
+	protected function get_offline_post_image_routes( $post ) {
+		$routes = array();
+		preg_match_all( '/< *img[^>]*src *= *["\']?([^"\']*)/i', $post->post_content, $matches );
+		if ( isset( $matches[1] ) && ! empty( $matches[1] ) ) {
+			foreach ( $matches[1] as $src ) {
+				$routes[] = array(
+					'url'      => $this->remove_url_scheme( $src ),
+					'revision' => $post->post_modified,
+				);
+			}
+		}
+
+		$featured_img_url = get_the_post_thumbnail_url( $post->ID );
+		if ( false !== $featured_img_url ) {
+			$routes[] = array(
+				'url'      => $this->remove_url_scheme( $featured_img_url ),
+				'revision' => $post->post_modified,
+			);
```

Comment:
Per above, we'll also need to calculate the image URLs for the `srcset` and `sizes`, as the `post-thumbnail` image may not actually be the size used on the offline page.